pkg/util/winutil/service.go

```diff
@@ -0,0 +1,232 @@
+	m, err := OpenSCManager(windows.SC_MANAGER_CONNECT)
+	if err != nil {
+		return nil, err
+	}
+	defer m.Disconnect()
+
+	s, err := OpenService(m, serviceName, windows.SERVICE_ENUMERATE_DEPENDENTS)
+	if err != nil {
+		return nil, fmt.Errorf("could not open service: %v", err)
+	}
+	defer s.Close()
+
+	deps, err := enumDependentServices(s.Handle, state)
+	if err != nil {
+		return nil, fmt.Errorf("could not enumerate dependent services: %v", err)
+	}
+	return deps, nil
+}
+
+// ServiceStatus reports information pertaining to enumerated services
+// only exported so binary.Read works
+type ServiceStatus struct {
+	DwServiceType             uint32
+	DwCurrentState            uint32
+	DwControlsAccepted        uint32
+	DwWin32ExitCode           uint32
+	DwServiceSpecificExitCode uint32
+	DwCheckPoint              uint32
+	DwWaitHint                uint32
+}
+
+// EnumServiceStatus complete enumerated service information
+// only exported so binary.Read works
+type EnumServiceStatus struct {
+	serviceName string
+	displayName string
+	status      ServiceStatus
+}
+
+type internalEnumServiceStatus struct {
+	ServiceName uint64 // offset from beginning of buffer
+	DisplayName uint64 // offset from beginning of buffer.
+	Status      ServiceStatus
+	Padding     uint32 // structure is qword aligned.
+
+}
+
+func enumDependentServices(h windows.Handle, state enumServiceState) (services []EnumServiceStatus, err error) {
+	services = make([]EnumServiceStatus, 0)
+	var bufsz uint32
+	var count uint32
+	_, _, err = procEnumDependentServices.Call(uintptr(h),
+		uintptr(state),
+		uintptr(0),
+		uintptr(0), // current buffer size is zero
+		uintptr(unsafe.Pointer(&bufsz)),
+		uintptr(unsafe.Pointer(&count)))
+
+	// success with a 0 buffer means no dependent services
+	if err == error(windows.ERROR_SUCCESS) {
+		err = nil
+		return
+	} else {
+		if err != error(windows.ERROR_MORE_DATA) {
+			log.Warnf("Error getting buffer %v", err)
+			return
+		}
+	}
+
+	servicearray := make([]uint8, bufsz) // TODO convert to uint16 and bufsz/2 to allow windows.UTF16toString call
+	ret, _, err := procEnumDependentServices.Call(uintptr(h),
+		uintptr(state),
+		uintptr(unsafe.Pointer(&servicearray[0])),
+		uintptr(bufsz),
+		uintptr(unsafe.Pointer(&bufsz)),
+		uintptr(unsafe.Pointer(&count)))
+	if ret == 0 {
+		log.Warnf("Error getting deps %d %v", int(ret), err)
+		return
+	}
+	// now get to parse out the C structure into go.
+	var ess internalEnumServiceStatus
+	baseptr := uintptr(unsafe.Pointer(&servicearray[0]))
+	buf := bytes.NewReader(servicearray)
+	for i := uint32(0); i < count; i++ {
+
+		err = binary.Read(buf, binary.LittleEndian, &ess)
+		if err != nil {
+			break
+		}
+
+		ess.ServiceName = ess.ServiceName - uint64(baseptr)
+		ess.DisplayName = ess.DisplayName - uint64(baseptr)
+		ss := EnumServiceStatus{serviceName: ConvertWindowsString(servicearray[ess.ServiceName:]),
```

Comment:
can you not pass `ess.ServiceName` instead of using the offset with `servicearray[ess.ServiceName:]`?